brayns/io/MeshLoader.cpp

```diff
@@ -333,6 +309,106 @@ void MeshLoader::_createMaterials(Scene& scene, const aiScene* aiScene,
+    auto& triangleMeshes = scene.getTriangleMeshes();
+    for (size_t m = 0; m < aiScene->mNumMeshes; ++m)
+    {
+        aiMesh* mesh = aiScene->mMeshes[m];
+        const size_t materialId =
+            _getMaterialId(mesh->mMaterialIndex, defaultMaterial);
+
+        auto& triangleMesh = triangleMeshes[materialId];
+
+        nbVertices += mesh->mNumVertices;
+        triangleMesh.vertices.reserve(nbVertices);
+        if (mesh->HasNormals())
+            triangleMesh.normals.reserve(nbVertices);
+        if (mesh->HasTextureCoords(0))
+            triangleMesh.textureCoordinates.reserve(nbVertices);
+        for (size_t i = 0; i < mesh->mNumVertices; ++i)
+        {
+            const auto& v = mesh->mVertices[i];
+            const Vector4f vertex =
+                transformation * Vector4f(v.x, v.y, v.z, 1.f);
+            const Vector3f transformedVertex = {vertex.x(), vertex.y(),
+                                                vertex.z()};
+            triangleMesh.vertices.push_back(transformedVertex);
+            scene.getWorldBounds().merge(transformedVertex);
+            if (mesh->HasNormals())
+            {
+                const auto& n = mesh->mNormals[i];
+                const Vector4f normal =
+                    transformation * Vector4f(n.x, n.y, n.z, 0.f);
+                const Vector3f transformedNormal = {normal.x(), normal.y(),
+                                                    normal.z()};
+                triangleMesh.normals.push_back(transformedNormal);
+            }
+
+            if (mesh->HasTextureCoords(0))
+            {
+                const auto& t = mesh->mTextureCoords[0][i];
+                const Vector2f texCoord(t.x, -t.y);
+                triangleMesh.textureCoordinates.push_back(texCoord);
+            }
+        }
+        bool nonTriangulatedFaces = false;
+        nbFaces += mesh->mNumFaces;
+        triangleMesh.indices.reserve(nbFaces);
+        if (_meshIndex.find(materialId) == _meshIndex.end())
+            _meshIndex[materialId] = 0;
+        const auto meshIndex = _meshIndex[materialId];
+        for (size_t f = 0; f < mesh->mNumFaces; ++f)
+        {
+            if (mesh->mFaces[f].mNumIndices == 3)
+            {
+                const Vector3ui ind =
+                    Vector3ui(meshIndex + mesh->mFaces[f].mIndices[0],
+                              meshIndex + mesh->mFaces[f].mIndices[1],
+                              meshIndex + mesh->mFaces[f].mIndices[2]);
+                triangleMesh.indices.push_back(ind);
+            }
+            else
+                nonTriangulatedFaces = true;
+        }
+        if (nonTriangulatedFaces)
+            BRAYNS_DEBUG
+                << "Some faces are not triangulated and have been removed"
+                << std::endl;
+
+        _meshIndex[materialId] += mesh->mNumVertices;
+
+        updateProgress("Post-processing mesh ...",
+                       50 + ((m + 1) * 50.f) / aiScene->mNumMeshes, 100);
```

Comment:
looks like the other half of the magic :-) I don't know if we could write a sort of proxy progress class that does the aggregation of the multiple steps.